accepted/future-releases/variance/feature-specification-use-site-variance.md

```diff
@@ -0,0 +1,453 @@
+
+It is a compile-time error if a type parameter declared by a static extension has a variance modifier.
+
+*Variance is not relevant to static extensions, because there is no notion of subsumption. Each usage will be a single call site, and the value of every type argument associated with an extension method invocation is statically known at the call site.*
+
+It is a compile-time error if a type parameter _X_ declared by a type alias has a variance modifier, unless it is `inout`; or unless it is `out` and the right hand side of the type alias has only covariant occurrences of _X_; or unless it is `in` and the right hand side of the type alias has only contravariant occurrences of _X_.
+
+*The variance for each type parameter of a type alias is restricted based on the body of the type alias. Explicit variance modifiers may be used to document how the type parameter is used on the right hand side, and they may be used to impose a variance on an unused type parameter.*
+
+We say that a type parameter _X_ of a type alias _F_ _is covariant/invariant/contravariant_ if it has the variance modifier `out`/`inout`/`in`, respectively. We say that it is _covariant/contravariant_ if it has no variance modifier, and it occurs only covariantly/contravariantly, respectively, on the right hand side of `=` in the type alias (*for an old-style type alias, rewrite it to the form using `=` and then check*). Otherwise (*when _X_ has no modifier, but occurs invariantly or both covariantly and contravariantly*), we say that _X_ _is invariant_.
+
+Let _D_ be the declaration of a class or mixin, and let _X_ be a type parameter declared by _D_.
+
+We say that _X_ _is covariant_ if it has no variance modifier or it has the variance modifier `out`; that it _is invariant_ if it has the variance modifier `inout`; and that it _is contravariant_ if it has the variance modifier `in`.
+
+If _X_ has the variance modifier `out` then it is a compile-time error for _X_ to occur in a non-covariant position in a member signature in the body of _D_, except that it is not an error if it occurs in a covariant position in the type annotation of a covariant formal parameter (*note that this is a contravariant position in the member signature as a whole*). *For instance, _X_ can not be the type of a method parameter (unless covariant), and it can not be the bound of a type parameter of a generic method.*
```

Comment:
We should probably make it explicit that this covers setters as well.  Same for line 202.

Reply:
We also should make it explicit that `covariant` on a field covers the setter, but *not* the getter.  So we have that:

```
class C<in T> {
   covariant C<T> f1; // Error, the C<T> appears in an invalid position in the getter
   final covariant C<T> f2; // Error as above
}

class D<out T> {
  covariant D<T> f1; // Ok: the getter is properly covariant, and the setter is covered by the covariant annotation.
}
```

cc @kallentu